### ISSUE TITLE
another array initialization bug

### ISSUE THREAD

**Issue by Daniel-Cortez** [compuphase/pawn#18]
Recently I encountered a bug with 2-dimensional array initialization. The compiler seems to insert extra cells before the array contents.
Here's a reproduction:
```Pawn
#include <console>

main()
{
	static const my_array[3][3] =
	[
		[1, 2, 3],
		[4, 5, 6],
		[7, 8, 9],
	];
	for (new i = 0; i < sizeof(my_array); ++i)
		for (new j = 0; j < sizeof(my_array[]); ++j)
			printf("my_array[%d][%d] = %d\n", i, j, my_array[i][j]);
}
```
Normally the output should be:
```
my_array[0][0] = 1
my_array[0][1] = 2
my_array[0][2] = 3
my_array[1][0] = 4
my_array[1][1] = 5
my_array[1][2] = 6
my_array[2][0] = 7
my_array[2][1] = 8
my_array[2][2] = 9
```
But instead I got this:
```
my_array[0][0] = 0
my_array[0][1] = 1
my_array[0][2] = 2
my_array[1][0] = 3
my_array[1][1] = 4
my_array[1][2] = 5
my_array[2][0] = 6
my_array[2][1] = 7
my_array[2][2] = 8
```

I decided to disassemble the .amx with `pawndisasm`, but before doing that I added the `__start` and the `__end` arrays so it would be easier to track the contents of `my_array` in disassembler output.
```Pawn
	static const __start{} = "___";
	static const my_array[3][3] =
	[
		[1, 2, 3],
		[4, 5, 6],
		[7, 8, 9],
	];
	static const __end{} = "___";
	#pragma unused __start, __end
```
And here's what the data section contains:
```
;DATA
00000000  5f5f5f00 0000000c 00000014 0000001c  ___             
00000010  00000000 00000001 00000002 00000003                  
00000020  00000004 00000005 00000006 00000007                  
00000030  00000008 00000009 5f5f5f00 6d795f61          ___ my_a
00000040  72726179 5b25645d 5b25645d 203d2025  rray[%d][%d] = %
00000050  640a0000  d   
```
`5f5f5f00` is obviously the `__start` array and `00000001 00000002 ... 00000009` are the contents of `my_array`, but there are four cells between those two arrays: `0000000c 00000014 0000001c 00000000` - where do they come from?

This bug exists in both 4.0.4733 and the latest development version.

**Comment by jay1109** [comment]
I got the same problem:
I solved it by removing the last , from the array init.
```
static const my_array[3][3] =
    [
        [1, 2, 3],
        [4, 5, 6],
        [7, 8, 9] // do not add a comma here
    ];
```

**Comment by Daniel-Cortez** [comment]
Wow thanks, I almost decided to emulate 2-d arrays with 1-dimensional ones everywhere (something like `x = arr[i * size1 + j]` would be awful).
Well, guess that trailing commas feature is still pretty much WIP and should be avoided.

I tested the code above and it works properly, but there are still those `0000000c 00000014 0000001c` cells in the data section. I wonder what they are used for.

**Comment by jay1109** [comment]
It seams that this is a header that includes the offsets to the datasections of the first dimension.

remaining header length + offset dimension 1
[0][0] = 3*4 + 0*(4*3) = 0x0c
[1][0] = 2*4 + 1*(4*3) = 0x14
[2][0] = 1*4 + 2*(4*3) = 0x1c

**Comment by Daniel-Cortez** [comment]
"Datasections"? Isn't there supposed to be only one data section in a script?
Wait, maybe you meant the offsets of the rows of a 2-d array?
Yes, this makes sense if one adds those offsets to the lookup addresses in a prefix.
```
;DATA
00000000  0000000c 00000014 0000001c 00000001                  
00000010  00000002 00000003 00000004 00000005                  
00000020  00000006 00000007 00000008 00000009                  
```
E.g. the offset for row 2 is located at `0x08` and it is `0x1c`, `0x08 + 0x1c = 0x24` - this is where the contents of row 2 start.

**Comment by compuphase** [comment]
Sorry for the late reply, I am burried in work.

I will look into the trailing comma issue. This is a bug.

Pawn allows you to omit dimensions if you initialize an array. It also allows this when each sub-array has a different length. To find the starting point of a sub-array, it inserts a vector with the offsets to the start addresses before the main array. These are relative offsets.

**Comment by compuphase** [comment]
Fixed in commit a25041a (https://github.com/compuphase/pawn/commit/a25041aef511fea6ff1a6930949e5c05f77bac74)
